core/psh/psh.c

```diff
@@ -972,7 +974,117 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, code;
+	const int maxinsz = 32;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1], shdw[128];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+	mode.c_lflag |= (ICANON | ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Get login from user */
+	memset(indata, '\0', sizeof(indata));
+
+	for (;;) {
+		if (insz < 0) {
+			fprintf(stdout, "Login: ");
+			fflush(stdout);
+			insz++;
+		}
+
+		if(read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (insz <= maxinsz && insz > 0) {
+				memset(indatasave, '\0', sizeof(indatasave));
+				memcpy(indatasave, indata, sizeof(indatasave));
+				memset(indata, '\0', sizeof(indata));
+				break;
+			}
+			else {
+				memset(indata, '\0', sizeof(indata));
+				insz = -1;
+				continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+			continue;
+		}
+	}
+
+	/* Get password from user */
+	memset(shdw, '\0', sizeof(shdw));
+	memset(indata, '\0', sizeof(indata));
+	mode.c_lflag &= ~(ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+	insz = -1;
+	for (;;) {
+		if (insz < 0) {
+			printf("Password: ");
+			fflush(stdout);
+			insz++;
+		}
+
+		if(read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+
+		if (c == '\n') {
+			memset(shdw, '\0', sizeof(shdw));
+			if (insz <= maxinsz) {
+				memcpy(shdw, crypt(indata, "ABCD"), sizeof(shdw)); 
+				memset(indata, '\0', sizeof(indata));
+				insz = 0;
+				break;
+			}
+			else {
+				memset(indata, '\0', sizeof(indata));
+				insz = -1;
+				break;
+			}
+		}
+		else if (isprint(c)) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+		}
+	}
+	fprintf(stderr, "\n");
+	mode.c_lflag |= (ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* check passed credentials */
+	code = -EACCES;
+	struct passwd *userdata;
+	userdata = getpwnam(indatasave);
+	if (userdata != NULL){
+		if (!strcmp(userdata->pw_name, indatasave) && !strcmp(userdata->pw_passwd, shdw)) {
+			code = 0;
```

Comment:
`return 1`, why complicate the flow?

You can remove the `code` variable altogether

Reply:
- Prefix added
- yes You are right. By not hardcoded i meant that the password shadow is not provided by in-code `#define` and can be project specific with possibility to simple change.
- definitely should not. Indents removed. Thanks

I did as You requested with `return` and `code` variable. Its now easier to see where functions terminates in case of success login, but each return now needs its own resetting terminal to original state. Still it is shorter than using `code` flag.

Reply:
if we would handle the terminal in input function, we could exit early without consequences :)